### ISSUE TITLE
Failing on learning PPO

### ISSUE THREAD

**Issue by carlorop** [tensorforce/tensorforce#845]
I am training a PPO agent using the following code, after initializing an instance of the environment in the variable env:

```python
from tensorforce import Agent

agent = Agent.create(agent='ppo', states={"type":"float","shape":(36,)},actions={"type":"float","shape":(2,), "min_value":0.0, "max_value":1.0}, batch_size=64, learning_rate=1e-3,memory=13000000,max_episode_timesteps=20000,exploration=0.05)

n_episodes = 100
max_t = 10000

for i_episode in range(1, n_episodes + 1):
    state = env.reset()
    previous_iteration_states = [state]
    for t in range(1, max_t + 1):
            action = agent.act(state.flatten(), deterministic=False, independent=False)
            action_det = agent.act(state.flatten(), deterministic=True, independent=True)
            print(action,action_det)
            next_state, reward, done, _ = env.step(action)
            print(f"reward={reward} in step # {env.current_step}\n")
            agent.observe(reward=reward,terminal=done)
            if done:
                break

```

The problem that I had was that in the evaluation all actions were `0.5`. For this reason I am printing the action and a deterministic version of the action.  After some thousands of iterations, we can see the following random chunk of the results:

```
...

[0.7872943 0.369286 ] [0.4999973  0.50000685]

reward=0.7446212214571036 in step # 8732

[0.5656449  0.34001142] [0.4999997 0.5000038]

reward=0.7447317732621439 in step # 8733

[0.42484254 0.33627918] [0.4999959  0.49999934]

reward=0.7448427867189571 in step # 8734

[0.805025   0.21069574] [0.49999836 0.5000003 ]

reward=0.7449537898659087 in step # 8735

[0.5224615  0.86530674] [0.49999952 0.4999961 ]

...

```

As can be seen, the actions during training are not 0.5 because of the random noise. It seems that the agent is not learning. Is there something wrong in the initialization of the agent? I have tried to solve this changing the network, the exploration and the update frequency.

## Solution

The problem was that PPO learns using Advantage Estimator over episodes. This implies that when a PPO agent is trained, it will not update the parameters until the end of the episode, which is specified by the variable done. In my case, the end of the training was specified by the maximum number of timesteps, which were less than the number of episodes needed to return `done=1`.  After changing it the agent is training again.

**Comment by ValentinGlaser** [comment]
I'm relatively new to RL and I'm wondering on what the learning depends besides the hyperparameters. (I am using **PPO**)
In my problem hyperparameter tuning has been done (with a good learning curve as a result) but I'm struggling to reproduce that learning curve (reward oszillates around same value). Therefore I'm suspecting that it is a general problem.
I didn't quite unterstood how the number of max_episode_timesteps play a role. 

- The hyperparameter _memory_ is not explicitly definied, could this be the source of the non-learning?
- I suspect my problem has nothing to do with the hyperparameters, so do you have any clues where my problem could lie? (Env definition, ...)

I know that there are endless possible sources of my problem, but I'm searching for clues. Maybe you got some. 
Thanks.

**Comment by AlexKuhnle** [comment]
- Learning can be very unstable in RL, so it may well be that the hyperparams you identified only rarely lead to good performance. Can you post your hyperparams?
- `max_episode_timesteps` is an important parameter, since it specifies the length of an episode. Moreover, handling episode terminals in the right way is important. Do you use `Environment.create(...)`? to instantiate your environment? How do you choose the `max_episode_timesteps` argument?
- `memory` is specified implicitly when using PPO.
- Environment details are often the reason for learning difficulties, if it's a custom environment.

**Comment by ValentinGlaser** [comment]
Thank you for your quick response.

To give you some context: I am currently working on my thesis and the hyperparamater tuning was done in the thesis before mine and I'm using the same.

**hyperparameters:**
```
agent = Agent.create(
    agent='ppo',
    environment=env_RL,  
    update_frequency = 20 ,
    batch_size=20,
    learning_rate = 0.0001,
    multi_step = 10,
    subsampling_fraction = 0.33,
    likelihood_ratio_clipping = 0.1,
    discount = 0.9,
    predict_terminal_values = False,
    state_preprocessing = "linear_normalization",
    exploration = 0.0,
    variable_noise = 0.0,
    l2_regularization = 0.01,
    entropy_regularization = 0.01,
    )
```

- yes,`Environment.create(...)` is used to create the  custom environment `env_RL`
```
env_RL = Environment.create(
        environment=CustomEnvironment, max_episode_timesteps=200000
    )
```

- `max_episode_timesteps` is choosen quite randomly to be honest. How do I choose it? There have to be enough timesteps that the agent can perform all the actions, right? So why can't I choose an very high number for the timesteps?
- the episode terminal is handled like this: (`env` is a `simpy` environment)
```
env.run(until=SIM_TIME)
terminal = True
```


_top_: my results
_bottom_: results of thesis before (same hyperparameters) -> desired
_left_: raw
_right_: smoothed
![image](https://user-images.githubusercontent.com/86845720/145825826-1898f832-7f6b-4cf7-b25d-df83cd561026.png)

**Comment by AlexKuhnle** [comment]
Let's start with `max_episode_timesteps` and termination. It sounds like your environment is stopped arbitrarily at a certain time, which means that your terminal is not a "failure" (as it  e.g. often is in games), but an arbitrary stop of simulation. This is a subtle but somewhat important distinction, and in Tensorforce these cases are handled best by implementing an environment which *never* terminates (since there is no true termination), and instead specify `max_episode_timesteps` as part of `Environment.create(...)` to be the cut-off point. If that doesn't work, for whatever reason, you can also return a `terminal = 2` instead of `terminal = True`.

In general, the length of an episode is quite important, particularly when using PPO, since updates happen on an episode-basis. I would guess episodes in most cases should be around 100 to maybe up to 10,000 steps long (which already feels quite long). If longer (e.g. if your episodes are really 200,000 timesteps long), the questions I would ask are: (a) is the "time resolution" too fine-grained, i.e. are individual timesteps only corresponding to tiny intervals? (b) is the simulation unnecessarily long, i.e. the agent already sees enough and can understand the environment at, say, 10% of the timesteps. It's not important for episodes to go on forever during training, the agent just needs to see enough to see relevant developments. Your relatively aggressive `discount` value of `0.9` seems to suggest that actually the temporal dependencies are rather short.

Hope that makes sense.

Finally, two comments on your hyperparameters: (a) worth experimenting with values `update_frequency < batch_size`. (b) worth experimenting with `1e-4 (yours). <= learning_rate <= 1e-2`

**Comment by ValentinGlaser** [comment]
Thank you for your detailed reply.

1. I want to run simulations with a simulation time of 1h, 8h, 16h and 24h. Right now it is implemented that `terminal = True` when this time is elapsed in an `SimPy` environment. So how would I specify `max_episode_timesteps` that its results in exactly e.g. 8h simulation time?

2. (a) yes, I think so. The explanation of `max_episode_timesteps` I got was the following: "maximum number of decisions that the RL agent can make during an episode of the simulation". Is this correct?

3. (b) no, I don't think so because I want to simulate the whole simulation time

4. general question: are the advantages or disadvantages of `act_observe` or `act_experience_update`? Is the a combination possible?

5. `act_observe`: when does the update takes place?

6. The agent maximizes the reward by default? Even if I got a negative reward?

**Comment by AlexKuhnle** [comment]
1. If that is not possible, you can also return `terminal = 2` ("simulation-abort terminal") instead of `terminal = True` ("game-over terminal").
2. Yes. And if the act intervals are too tiny, that could be a problem.
3. That however doesn't mean that it's the best for the agent to train and learn from, if the episodes are too long.
4. You should use `act()` and `observe()`.
5. The update implicitly takes place as specified in the `update` argument (Tensorforce agent), or after `batch_size` and with frequency `update_frequency` epochs/timesteps (epochs in the case of PPO). `observe()` returns a number which gives the number of updates performed as part of this observe steps (usually 0, sometimes 1).
6. Yes, so if your rewards are negative, it will try to get them as close to zero as possible.

**Comment by ValentinGlaser** [comment]
I did get a learning curve with the standard hyperparameters! But I have a few further questions for my understanding.

2. How can those decisions the agent makes be counted? With `agent.act()` or with `environment.execute()`?

**Comment by AlexKuhnle** [comment]
Either should work, since every `act()` is followed by an `execute()`. It can also make sense to do the counting inside the `Environment` implementation, to separate the agent-env interaction code from any environment-specific logging.

**Comment by lazyaa** [comment]
I'm relatively new to RL ，and I wonder if the action changes during an episode of  PPO training? The problem I have is that no matter how much I modify the hyperparameters, the actions are always the same during the run of a single episode, and when I train multiple episodes, there is only one action in each episode, who can help me?
